### PR TITLE
Add comopnent to create an activity open trace based on a sequence of events

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

This is a component that takes a set of input events and creates the appropriate activity open trace. It's not hooked up to anything so no traces are going to be logged when this is merged, but it's sizeable enough that I want it reviewed separately, as its logic can be validated via tests.

## Testing
Unit tests verifying this components works properly if given the right input in the right order.

### CODE REVIEW

**Comment by bidetofevil** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1319?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1320** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1320?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1319** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1319?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1318** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1318?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1317** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1125** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1125?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1124** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1124?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1123** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1123?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-api/api/embrace-android-api.api
Comment (bidetofevil) [reply embrace-android-api/api/embrace-android-api.api:182]: I should add doc to more clearly articulate the design - I was going to do that before submitting this for review, but (I thought) the PR messiness with the trunk change had forced my hand.

So, the idea is that the set of events to create a trace is a super set of the lifecycle events and spans multiple invocations of activity opening. And how they are fired for the lifecycle events also depends on the Android version. This set of events effectively normalizes all those difference and combines both the lifecycle and non-lifecycle events (i.e. the render stuff). 

The listener for these events will be responsible for starting and stoping spans, so they only have to be aware of these, instead of all the messiness of what this intends to encapsulate and normalize. This design will make each component more easily testable, and tease out the logic and complexity.

File: embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt
Comment (bidetofevil) [reply embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt:17]: This is when one activity's `onStop()` is called without another's `onStart()` is called. It is meant to denote the app going into the background so another activity has not been started.

We keep track of this in order to properly gauge the start time of the next activity, as I wanted to account for the time between `onPause` and `onCreate` between two activities and have it as part of the open time for the latter if we are doing a foreground transition. Obviously we don't want to track this if the app was previously backgrounded.

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:18]: I need to take a look again. It was correct at some point in time but I'm not sure I've reviewed it

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:28]: The UI frame thing? That's when the activity opening ends. The start is when it comes from the background or when the previous activity is done. The former is when we can detect onCreate (so onActivityPreCreated() if possible, and just onActivityCreated() if it's not depending on API version), while the latter is onActivityPostPaused() if possible, and just onActivityPaused().

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:34]: Yeah. The assumption that one activity's pause is when the next activity starts to open will be broken. Without this assumption, we wouldn't really know the when activity's opening is triggered, unless we we can somehow get that info from the intent?

I would like to not have to make this assumption, but if we do, we should call it out. I don't know how multi-screen windowing works with respect to the activity lifecycle callbacks - I feel like a lot of things we have that uses that API will be broken if this assumption is incorrect.

So we should really look into that 😅

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:40]: Activities are never obfuscated. That's in the default keep rules. I hadn't realized this until I started looking at this stuff

File: embrace-android-api/api/embrace-android-api.api
Comment (fractalwrench) [reply embrace-android-api/api/embrace-android-api.api:182]: `embrace-android-features` or `embrace-android-core` will hide it from prying eyes

File: embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt
Comment (fractalwrench) [reply embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt:10]: An activity can be passed an intent and alter its behavior depending on the bundle inside - I don't think it's safe to assume 1 activity type == only 1 instance at a time. I've certainly written apps in the past that host different fragments etc based on intent args, so there may be apps in the wild that made similar decisions

Comment (bidetofevil) [reply embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt:10]: Got it. So the case you're concerned about is FooActivity opening another instance of FooActivity?

I /believe/ this case is handled correctly if the first instance completes its lifecycle before starting another instance of the same activity. I will write a test to verify this.

The case I was talking about is regarding a new Activity instance beginning to initialize before the previous one is paused, in which case the unfinished instance, regardless of type, will be assumed to be abandoned. 

For the startup trace, I had initially made the incorrect assumption that if the type is the same that it must be the same instance, which is a subset of the case you're talking about and cannot be assumed. This is why I'm using the hash code rather than the type, to prevent the case where a new implementation of FooActivity is started before the previous instance has completed

File: embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt
Comment (fractalwrench) [reply embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt:17]: Got it. So this is equivalent to `ProcessStateService#onBackground`

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (fractalwrench) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:28]: I'm struggling to visualise what the trace will look like in a dashboard but think that's probably just because this PR is a small part of a bigger picture.

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (fractalwrench) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:34]: We could probably just document that it's not explicitly supported I guess? It's not a common use-case so we might be building something no/few folks want.

File: embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt
Comment (bidetofevil) [reply embrace-android-api/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenEvents.kt]: Yeah, except I think this will fire for an activity even if it opened another activity, whereas I think the `LifecycleEventObserver` only fires `onStop` if it doesn't detect that another activity is being opened?

Hmm. Makes me think whether I should hook into this instead of the activity itself to record these traces....

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:40]: Eliab put together a really comprehensive outline here: https://www.notion.so/embraceio/Tracing-Auto-Instrumentation-1603dd140a8641ceb9fc8f0a1c63f3ef?pvs=4

There are probably differences between this impl and what is spec'd out, but largely this is what we want to achieve

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt:47]: Yeah, if it's even possible. I'll confirm or call out.

File: embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt
Comment (bidetofevil) [reply embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/OpenTraceEmitter.kt]: Reviewed. Looks right.

**Review by bidetofevil — COMMENTED** [review]
Added an explanation of the overall design and posted it on Slack.

File: embrace-android-features/src/test/java/io/embrace/android/embracesdk/internal/capture/activity/UiLoadTraceEmitterTest.kt
Comment (fractalwrench) [inline embrace-android-features/src/test/java/io/embrace/android/embracesdk/internal/capture/activity/UiLoadTraceEmitterTest.kt:66]: Test cases with the ignore annotation should be updated to run before merging, or can be removed if they're no longer relevant

**Review by fractalwrench — APPROVED** [review]
LGTM once a couple of ignored test cases are resolved

File: embrace-android-features/src/test/java/io/embrace/android/embracesdk/internal/capture/activity/UiLoadTraceEmitterTest.kt
Comment (bidetofevil) [reply embrace-android-features/src/test/java/io/embrace/android/embracesdk/internal/capture/activity/UiLoadTraceEmitterTest.kt:66]: Ah, I think they work - so I'll either un-ignore or fix them

**Comment by bidetofevil** [comment]
### Merge activity

* **Nov 29, 1:21 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1319).

**Pull request merged**
